lib/src/barback/asset_environment.dart

```diff
@@ -167,21 +165,51 @@ class AssetEnvironment {
   /// go to barback immediately.
   Set<AssetId> _modifiedSources;
 
+  /// The compiler mode for this environment.
+  final Compiler compiler;
+
   AssetEnvironment._(this.graph, this.barback, this.mode, this._watcherType,
-      this._hostname, this._basePort, this.environmentConstants);
+      this._hostname, this._basePort, this.environmentConstants, this.compiler);
 
-  /// Gets the built-in [Transformer]s that should be added to [package].
+  /// Gets the built-in [Transformer]s or [AggregateTransformer]s that should be
+  /// added to [package].
   ///
   /// Returns `null` if there are none.
-  Iterable<Transformer> getBuiltInTransformers(Package package) {
-    // Built-in transformers only apply to the root package.
-    if (package.name != rootPackage.name) return null;
+  Iterable<Set> getBuiltInTransformers(Package package) {
+    var transformers = <Set>[];
 
-    // The built-in transformers are for dart2js and forwarding assets around
-    // dart2js.
-    if (_builtInTransformers.isEmpty) return null;
+    if (compiler == Compiler.dartDevc) {
+      throw new UnimplementedError(
+          'The dartdevc compiler is not yet supported.');
+    }
+
+    // These transformers are just for the root package.
+    if (package.name == rootPackage.name) {
+      switch (compiler) {
+        case Compiler.dart2Js:
+          // If the entrypoint package manually configures the dart2js
+          // transformer, don't include it in the built-in transformer list.
+          //
+          // TODO(nweiz): if/when we support more built-in transformers, make
+          // this more general.
+          var containsDart2JS = graph.entrypoint.root.pubspec.transformers.any(
+              (transformers) => transformers
+                  .any((config) => config.id.package == '\$dart2js'));
+
+          if (!containsDart2JS && compiler == Compiler.dart2Js) {
+            transformers.add([
+              new Dart2JSTransformer(this, mode),
+              new DartForwardingTransformer(),
+            ].toSet());
+          }
+          break;
+        default:
+          break;
+      }
+    }
 
-    return _builtInTransformers;
+    if (transformers.isEmpty) return null;
```

Comment:
You add an explicit empty->null here, and then later have to check for null before adding. How about just returning the empty collection?